contracts/root/lib/ChildChainVerifier.sol

```diff
@@ -0,0 +1,167 @@
+    );
+
+    require(txList[5].toBytes().length == 36, "MALFORMED_WITHDRAW_TX");
+    // check withdraw function signature
+    require(
+      // keccak256('withdraw(uint256)') = 0x2e1a7d4d
+      BytesLib.toBytes4(BytesLib.slice(txList[5].toBytes(), 0, 4)) == 0x2e1a7d4d,
+      "WITHDRAW_SIGNATURE_NOT_FOUND"
+    );
+
+    require(registry.isERC721(rootToken) || amountOrTokenId > 0);
+    require(amountOrTokenId == BytesLib.toUint(txList[5].toBytes(), 4));
+
+    // Make sure this tx is the value on the path via a MerklePatricia proof
+    // @todo This might be possible to remove
+    require(
+      MerklePatriciaProof.verify(txBytes, path, txProof, txRoot),
+      "INVALID_TX_MERKLE_PROOF"
+    );
+
+    // raw tx
+    bytes[] memory rawTx = new bytes[](9);
+    for (uint8 i = 0; i <= 5; i++) {
+      rawTx[i] = txList[i].toBytes();
+    }
+    rawTx[4] = hex"";
+    rawTx[6] = networkId;
+    rawTx[7] = hex"";
+    rawTx[8] = hex"";
+
+    // recover sender from v, r and s
+    require(
+      sender == ecrecover(
+        keccak256(RLPEncode.encodeList(rawTx)),
+        Common.getV(txList[6].toBytes(), Common.toUint8(networkId)),
+        bytes32(txList[7].toUint()),
+        bytes32(txList[8].toUint())
+      )
+    );
+  }
+
+  function processWithdrawTransferTx(bytes memory txBytes, address _registry)
+    internal
+    view
+    returns (address rootToken)
+  {
+    // check transaction
+    RLPReader.RLPItem[] memory items = txBytes.toRlpItem().toList();
+    require(items.length == 9);
+
+    // check rootToken is valid
+    Registry registry = Registry(_registry);
+    rootToken = registry.childToRootToken(items[3].toAddress());
+    require(rootToken != address(0));
+    // check if transaction is transfer tx
+    // <4 bytes transfer event,address (32 bytes),amountOrTokenId (32 bytes)>
+    bytes4 transferSIG = BytesLib.toBytes4(BytesLib.slice(items[5].toBytes(), 0, 4));
+    require(
+      // keccak256('transfer(address,uint256)') = 0xa9059cbb
+      transferSIG == 0xa9059cbb ||
+      // keccak256('transferFrom(address,adress,uint256)') = 0x23b872dd
+      (registry.isERC721(rootToken) && transferSIG == 0x23b872dd));
+  }
+
+  function processWithdrawTransferReceipt(bytes memory receiptBytes, address sender, address _registry)
+    internal
+    view
+    returns (uint256 totalBalance, uint8 oIndex)
+  {
+    RLPReader.RLPItem[] memory items = receiptBytes.toRlpItem().toList();
+    require(items.length == 4);
+
+    // retrieve LogTransfer event (UTXO <amount, input1, input2, output1, output2>)
+    items = items[3].toList()[1].toList();
+
+    // get topics
+    RLPReader.RLPItem[] memory topics = items[1].toList();
+
+    // get from/to addresses from topics
+    address from = BytesLib.toAddress(topics[2].toBytes(), 12);
+    address to = BytesLib.toAddress(topics[3].toBytes(), 12);
+
+    Registry registry = Registry(_registry);
+    if (registry.isERC721(address(topics[1].toUint()))) {
```

Comment:
I think previous logic was better in case of ERC721 it was termination early 😛.

Reply:
Good catch.!